### PR TITLE
Enhanced User Debugging

### LINKED ISSUE #2919: Error Email or SMS Notifications for User Debug.
One could use our User Debug turned on via Globals Logging that reports just PHP errors and/or console errors to the display by adding an option like Notify Administrators Only and then intercept errors in utility.js, qualify error for where to report then do an ajax to send email/sms (using our email facilities) to a list from users or created lists(notify on errors email/sms etc) in Lists. 

Could also include server side logs if necessary.

Would actually be a nice feature add where administrators could be aware of issues w/o user having to report from display notifications.

### PR DESCRIPTION
**Pull request by tywrenn**
Resolves #2919 

To-do List:

- [x] Add user-based logging (this helps me out a lot)
- [ ] Create a configurable interface where we can get notified on certain errors
- [ ] Improve our current logging system

@sjpadgett Dunno if you have anything to add here

### CODE REVIEW

File: library/globals.inc.php [hunk 1]
```diff
@@ -2613,14 +2630,15 @@ function gblTimeZones()
             xl('User Debugging Options'),
             array(
                 '0' => xl('None'),
-                '1' => xl('Display Window Errors Only'),
-                '2' => xl('Display Application Errors Only'),
-                '3' => xl('All'),
+                '1' => xl('Display Window (JavaScript) Errors Only'),
+                '2' => xl('Display Only PHP Inline Application Warnings'),
+                '3' => xl('Display All PHP Inline Logs Only'),
+                '4' => xl('Display All'),
```
Comment (sjpadgett) [inline library/globals.inc.php]: We don't want to display php warnings to display, just errors. Also, no need to log anything or try to fetch logs from environment(location issues).
I think best option here would be to leave this select as it is then add a boolean/or new selection for admin notifications. This way the user can still have their debug notification preference but admin gets their way too...

File: interface/globals.php [hunk 2]
```diff
@@ -604,8 +606,13 @@ function strterm($string, $length)
 
 // turn off PHP compatibility warnings
 ini_set("session.bug_compat_warn", "off");
+
 // user debug mode
-if ((int) $GLOBALS['user_debug'] > 1) {
-    error_reporting(error_reporting() & ~E_WARNING & ~E_NOTICE & ~E_USER_WARNING);
+if ($GLOBALS['user_debug'] > 1) {
+    if ($GLOBALS['user_debug'] >= 3) {
+        error_reporting(E_ALL);
```
[the remark this answers is not in the document]
Comment (tywrenn) [reply interface/globals.php]: This is more for long-term fixing with PHP. Most notices in OpenEMR can be cleared with just adding a detection of if, for example, we are posting to a form which would just require an if statement to check if we posted before grabbing stuff from the post index variables. Minimizing notices can help theoretically speed up PHP and optimize OpenEMR (and also prevent errors from happening long-term as notices are technically errors that allow code to execute anyway even if the error happened). There are differing opinions of if clearing up notices actually makes things run better, yet in my class that I took on PHP I was told to not have any notices at all when I code things. Putting this in will allow people to fix those if they want to (not that they have to).

File: library/globals.inc.php (hunk 1 above)
Comment (tywrenn) [reply library/globals.inc.php]: After Googling it appears I named what that setting actually does incorrectly woops. It should've stated it only displays errors

File: interface/globals.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjpadgett) [reply interface/globals.php]: > I was told to not have any notices at all when I code things

Granted but remember that most these admin are going to ignore notices. If it doesn't stop workflow they will concentrate on more important tasks albeit, annoyed from all the emails they are receiving.:)
If this is important to you then may I suggest allowing user to select the level of errors they wish to be notified about.

Also, i'm very interested to see how you will capture the php errors to report. Interesting indeed.

File: library/globals.inc.php (hunk 1 above)
Comment (sjpadgett) [reply library/globals.inc.php]: Ah Googling, our newest verb!

**Pull request closed without merging**